FG fails entire string if a character is used that does not exist in unicode

This is due to an assert deep inside the bidi module:

```
Traceback for <Task finished name='Task-15' coro=<FGMainWindowController.textEntryChangedCallback() done, defined at /Users/just/code/git/fontgoggles/Lib/fontgoggles/mac/mainWindow.py:609> exception=AssertionError(' not allowed here')> (most recent call last):
  File "/Users/just/code/git/fontgoggles/Lib/fontgoggles/mac/mainWindow.py", line 615, in textEntryChangedCallback
    self.textInfo = TextInfo(sender.get())
  File "/Users/just/code/git/fontgoggles/Lib/fontgoggles/misc/textInfo.py", line 10, in __init__
    self.text = text
  File "/Users/just/code/git/fontgoggles/Lib/fontgoggles/misc/textInfo.py", line 23, in text
    self._segments, self.baseLevel = textSegments(text)
  File "/Users/just/code/git/fontgoggles/Lib/fontgoggles/misc/segmenting.py", line 25, in textSegments
    storage = getBiDiInfo(txt)
  File "/Users/just/code/git/fontgoggles/Lib/fontgoggles/misc/segmenting.py", line 147, in getBiDiInfo
    resolve_implicit_levels(storage, debug)
  File "/Users/just/code/git/fontgoggles/venv39/lib/python3.9/site-packages/bidi/algorithm.py", line 468, in resolve_implicit_levels
    assert _ch['type'] in ('L', 'R', 'EN', 'AN'),\
AssertionError:  not allowed here
```

`_ch['type']` is an empty string, which is returned by `unicodedata.bidirectional()` for such a character (for example U+1FF0).

--- a/Lib/fontgoggles/misc/segmenting.py
+++ b/Lib/fontgoggles/misc/segmenting.py
@@ -129,6 +129,7 @@ def getBiDiInfo(text, *, upper_is_rtl=False, base_dir=None, debug=False):
     storage['base_dir'] = ('L', 'R')[base_level]
 
     get_embedding_levels(text, storage, upper_is_rtl, debug)
+    fix_bidi_type_for_unknown_chars(storage)
     assert len(text) == len(storage["chars"])
     for index, (ch, chInfo) in enumerate(zip(text, storage["chars"])):
         assert ch == chInfo["ch"]
@@ -141,3 +142,9 @@ def getBiDiInfo(text, *, upper_is_rtl=False, base_dir=None, debug=False):
     reorder_resolved_levels(storage, debug)
 
     return storage
+
+
+def fix_bidi_type_for_unknown_chars(storage):
+    for _ch in storage['chars']:
+        if _ch['type'] == '':
+            _ch['type'] = 'L'

--- a/Lib/fontgoggles/misc/segmenting.py
+++ b/Lib/fontgoggles/misc/segmenting.py
@@ -145,6 +145,9 @@ def getBiDiInfo(text, *, upper_is_rtl=False, base_dir=None, debug=False):
 
 
 def fix_bidi_type_for_unknown_chars(storage):
+    """Set any bidi type of '' (symptom of a character not known by unicode)
+    to 'L', to prevent the other bidi code to fail (issue 313).
+    """
     for _ch in storage['chars']:
         if _ch['type'] == '':
             _ch['type'] = 'L'

--- a/Tests/test_segmenting.py
+++ b/Tests/test_segmenting.py
@@ -134,6 +134,7 @@ def test_detectScript(testString, expectedScripts):
 
 testDataTextSegments = [
     ("a", 0, [("a", "Latn", 0, 0)]),
+    ("\u1FF0", 0, [("\u1FF0", "Zzzz", 0, 0)]),  # test char should *not* be defined in Unicode: issue #313
     ("\u0627", 1, [("\u0627", "Arab", 1, 0)]),
     ("a\u0627", 0, [("a", "Latn", 0, 0), ("\u0627", "Arab", 1, 1)]),
     ("\u0627123", 1, [("\u0627", "Arab", 1, 0), ("123", "Arab", 2, 1)]),

